File: 프로그래머스/0/120812. 최빈값 구하기/최빈값 구하기.py
def solution(array):
    if len(array) == 0:
        return -1

    # 빈도 계산을 위한 딕셔너리 준비
    frequency_dict = {}

    # 배열을 순회하며 빈도를 계산
    for num in array:
        if num in frequency_dict:
            frequency_dict[num] += 1
        else:
            frequency_dict[num] = 1

    # 최대 빈도 및 최빈값 찾기
    freq_list = list(frequency_dict.items())
    max_freq = freq_list[0][1]
    max_key = freq_list[0][0]
    is_unique = True

    for i in range(1, len(freq_list)):
        key, freq = freq_list[i]

        if freq > max_freq:
            max_freq = freq
            max_key = key
            is_unique = True
        elif freq == max_freq:
            is_unique = False

    if is_unique:
        answer = max_key
    else:
        answer = -1

    return answer


print(solution([1, 2, 3, 3, 3, 4])) # 3
print(solution([1, 1, 2, 2]))       # -1
print(solution([1]))                # 1

File: 프로그래머스/0/120812. 최빈값 구하기/최빈값 구하기.java
class Solution {
    public int solution(int[] array) {
        if (array.length == 0) {
            return -1;
        }

        // 고유 숫자와 빈도를 저장할 배열
        int[] uniqueNums = new int[array.length];
        int[] frequencies = new int[array.length];
        int uniqueCount = 0;

        // 배열을 순회하며 빈도를 계산
        for (int num : array) {
            boolean found = false;
            for (int i = 0; i < uniqueCount; i++) {
                if (uniqueNums[i] == num) {
                    frequencies[i]++;
                    found = true;
                    break;
                }
            }
            if (!found) {
                uniqueNums[uniqueCount] = num;
                frequencies[uniqueCount] = 1;
                uniqueCount++;
            }
        }

        // 최대 빈도 및 최빈값 찾기
        int maxFreq = 0;
        int mode = -1;
        boolean isUnique = true;

        for (int i = 0; i < uniqueCount; i++) {
            if (frequencies[i] > maxFreq) {
                maxFreq = frequencies[i];
                mode = uniqueNums[i];
                isUnique = true;
            } else if (frequencies[i] == maxFreq) {
                isUnique = false;
            }
        }

        if (isUnique) {
            return mode;
        }
        return -1;
    }
}